File: tools/allstar-disc-extractor/confidence.py
"""Per-disc confidence scoring for archive quality."""

from __future__ import annotations

from typing import Any

GEOMETRY_TOLERANCE_DEG = 1.0


def compute_ocr_confidence(
    *,
    labeled: int,
    wedge_count: int,
    player: str,
    position: str,
    ocr_status: str,
) -> float:
    total = max(wedge_count, 1)
    wedge_ratio = min(labeled / total, 1.0)
    player_score = 1.0 if len(player.strip()) >= 3 else 0.25
    position_score = 1.0 if position.strip() else 0.4
    base = wedge_ratio * 0.55 + player_score * 0.3 + position_score * 0.15
    if ocr_status == "partial":
        base *= 0.75
    return round(max(0.0, min(100.0, base * 100)), 1)


def compute_geometry_confidence(*, degrees_sum: float) -> float:
    if degrees_sum <= 0:
        return 0.0
    error = abs(degrees_sum - 360.0)
    if error <= GEOMETRY_TOLERANCE_DEG:
        return 100.0
    if error <= 2.0:
        return 88.0
    if error <= 5.0:
        return 65.0
    if degrees_sum >= 200:
        return 40.0
    return 20.0


def compute_archive_confidence(*, ocr: float, geometry: float, validation_status: str) -> float:
    validation_score = 100.0 if validation_status == "validated" else 55.0 if validation_status == "warning" else 25.0
    score = ocr * 0.4 + geometry * 0.4 + validation_score * 0.2
    return round(max(0.0, min(100.0, score)), 1)


def trust_level(*, archive: float, ocr: float, geometry: float) -> str:
    if archive >= 85 and ocr >= 75 and geometry >= 85:
        return "trusted"
    if archive >= 55 or (ocr >= 45 and geometry >= 45):
        return "review_recommended"
    return "review_required"


def build_confidence_block(record: dict[str, Any]) -> dict[str, Any]:
    degrees = record.get("degrees") or {}
    labeled = sum(1 for v in degrees.values() if float(v or 0) > 0)
    wedge_count = int(record.get("wedgeCount") or 16)
    degrees_sum = float(record.get("degreesSum") or 0)
    ocr_status = str(record.get("ocrStatus") or "pending")
    validation_status = str(record.get("validationStatus") or "pending")

    ocr = compute_ocr_confidence(
        labeled=labeled,
        wedge_count=wedge_count,
        player=str(record.get("player") or ""),
        position=str(record.get("position") or ""),
        ocr_status=ocr_status,
    )
    geometry = compute_geometry_confidence(degrees_sum=degrees_sum)
    archive = compute_archive_confidence(ocr=ocr, geometry=geometry, validation_status=validation_status)
    trust = trust_level(archive=archive, ocr=ocr, geometry=geometry)

    return {
        "ocrConfidence": ocr,
        "geometryConfidence": geometry,
        "archiveConfidence": archive,
        "trustLevel": trust,
    }


def confidence_from_result(result, *, geometry_ok: bool, ocr_status: str, validation_status: str) -> dict[str, Any]:
    labeled = sum(1 for w in result.wedges if w.label is not None)
    wedge_count = len(result.wedges) or 16
    degrees_sum = sum(result.totals().values())
    ocr = compute_ocr_confidence(
        labeled=labeled,
        wedge_count=wedge_count,
        player=result.player,
        position=result.position,
        ocr_status=ocr_status,
    )
    geometry = compute_geometry_confidence(degrees_sum=degrees_sum)
    archive = compute_archive_confidence(ocr=ocr, geometry=geometry, validation_status=validation_status)
    trust = trust_level(archive=archive, ocr=ocr, geometry=geometry)
    return {
        "ocrConfidence": ocr,
        "geometryConfidence": geometry,
        "archiveConfidence": archive,
        "trustLevel": trust,
    }

File: lib/ops/allstar/confidence.ts
import type { AllStarArchiveRecord } from "./types";

export type TrustLevel = "trusted" | "review_recommended" | "review_required";

export type DiscConfidence = {
  ocrConfidence: number;
  geometryConfidence: number;
  archiveConfidence: number;
  trustLevel: TrustLevel;
};

export function trustFromScores(archive: number, ocr: number, geometry: number): TrustLevel {
  if (archive >= 85 && ocr >= 75 && geometry >= 85) return "trusted";
  if (archive >= 55 || (ocr >= 45 && geometry >= 45)) return "review_recommended";
  return "review_required";
}

export function computeConfidenceFromArchive(record: AllStarArchiveRecord): DiscConfidence {
  if (
    record.ocrConfidence != null &&
    record.geometryConfidence != null &&
    record.archiveConfidence != null &&
    record.trustLevel
  ) {
    return {
      ocrConfidence: record.ocrConfidence,
      geometryConfidence: record.geometryConfidence,
      archiveConfidence: record.archiveConfidence,
      trustLevel: record.trustLevel,
    };
  }

  const degrees = record.degrees ?? {};
  const labeled = Object.values(degrees).filter((v) => v > 0).length;
  const wedgeCount = record.wedgeCount ?? 16;
  const wedgeRatio = Math.min(labeled / Math.max(wedgeCount, 1), 1);
  const playerScore = record.player.trim().length >= 3 ? 1 : 0.25;
  const positionScore = record.position.trim() ? 1 : 0.4;
  let ocr = wedgeRatio * 0.55 + playerScore * 0.3 + positionScore * 0.15;
  if (record.ocrStatus === "partial") ocr *= 0.75;
  const ocrConfidence = Math.round(Math.max(0, Math.min(100, ocr * 100)) * 10) / 10;

  const sum = record.degreesSum ?? 0;
  let geometryConfidence = 0;
  if (sum > 0) {
    const error = Math.abs(sum - 360);
    if (error <= 1) geometryConfidence = 100;
    else if (error <= 2) geometryConfidence = 88;
    else if (error <= 5) geometryConfidence = 65;
    else if (sum >= 200) geometryConfidence = 40;
    else geometryConfidence = 20;
  }

  const validationScore =
    record.validationStatus === "validated" ? 100 : record.validationStatus === "warning" ? 55 : 25;
  const archiveConfidence =
    Math.round((ocrConfidence * 0.4 + geometryConfidence * 0.4 + validationScore * 0.2) * 10) / 10;

  return {
    ocrConfidence,
    geometryConfidence,
    archiveConfidence,
    trustLevel: trustFromScores(archiveConfidence, ocrConfidence, geometryConfidence),
  };
}

export function trustLabel(level: TrustLevel): string {
  if (level === "trusted") return "Trusted";
  if (level === "review_recommended") return "Review recommended";
  return "Review required";
}

export function trustTone(level: TrustLevel): "ok" | "warn" | "pending" {
  if (level === "trusted") return "ok";
  if (level === "review_recommended") return "warn";
  return "pending";
}
